app/data_providers/unleash_provider.py

```diff
@@ -0,0 +1,91 @@
+from typing import List
+
+from aws_xray_sdk.core import xray_recorder
+
+from app.data_providers.PocketGraphClientSession import PocketGraphClientSession
+from app.config import ENV, ENV_PROD
+from app.models.unleash_assignment import UnleashAssignmentModel
+from app.models.user import User
+
+
+class UnleashConfig:
+    APP_NAME = 'recommendation-api'
+    ENVIRONMENT = 'prod' if ENV == ENV_PROD else 'alpha'  # Must be prod, beta, or alpha (see UnleashEnvironment)
+
+
+class UnleashError(Exception):
+    pass
+
+
+class UnleashProvider:
+    """
+    Implements getting A/B test assignments from Unleash
+
+    Currently, the only supported Unleash strategies are 'Gradual rollout' and 'UserIDs', because we do not pass
+    any additional context to unleash beyond the `hashed_user_id` and `hashed_guid`. Pocket has created several custom
+    Unleash strategies that use additional context. For example the 'accountAge' strategy requires
+    `context.properties.accountCreatedAt` to be set. When required context for a strategy is not provided, then
+    AbTestAssignmentModel.assigned field will be `False`, meaning the user is ineligible for the experiment.
+    We should take care to only use assignments for which the required context is passed in.
+    @see https://github.com/Pocket/feature-flags/tree/main/src/unleashClient/strategy
+    """
+
+    def __init__(self, pocket_graph_client_session: PocketGraphClientSession, unleash_config: UnleashConfig):
+        self.pocket_graph_client_session = pocket_graph_client_session
+        self.unleash_config = unleash_config
+
+    @xray_recorder.capture_async('data_providers.UnleashProvider.get_assignments')
+    async def get_assignments(self, names: List[str], user: User) -> List[UnleashAssignmentModel]:
```

Comment:
I assume it will be called from the main methods of the API to determine the user's AB group. When an external service is called from tier1 API usually it's a good idea to use retries and fail on timeout and also provide the fallback. https://github.com/jd/tenacity is a good library for retries that can allow doing this pretty easily.

Failing of getting the AB group should not fail the request, so this method should provide a default alternative if the request is failed (in this case probably just an empty list).

Reply:
You raise an interesting point here. In the past we've chosen to use the [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) pattern to make it easy to notice what is wrong when a system is failing. Maybe it is time to be more fault-tolerant? Because this would be a shift in design, is it OK if we discuss this in retro, and keep it out-of-scope of this PR? I've already added it to [the agenda](https://docs.google.com/document/d/1RITPzyoXLFGI-eXwp7gicoSX4EBcXD5OodlgtJUW6_c/edit#heading=h.bv7ou16dveny) so we don't lose track.